### PR TITLE
[4.6][BLOCKER]CLOUDSTACK-8763: Resolved POD/ZONE deletion failure.

### PR DESCRIPTION
**Pull request by borisroman**
Instead of having both checkIfPodIsDeletable() and checkIfZoneIsDeletable have there own SQL query, I've refactored them so they use DAO SQL Queries.

This resolves the SQL Exception thrown by both classes.

Test to confirm working order:
- deploy ACS
- Add zones / pods. -> Try to delete without resources. -> Direct success.
- Add resources to zones / pods. -> Try to delete with resources in the pod / zone. -> Correct exception thrown. (Error message is why it cannot remove the zone / pod. IE: There is still a hosts or vm or .... )

### CODE REVIEW

File: engine/schema/src/com/cloud/network/dao/IPAddressDaoImpl.java [hunk 1]
```diff
@@ -282,6 +289,14 @@ public IPAddressVO findByVmIp(String vmIp) {
     }
 
     @Override
+    public int countIPs(long dcId, boolean onlyCountAllocated) {
+        SearchCriteria<Integer> sc = onlyCountAllocated ? AllocatedIpCount.create() : AllIpCount.create();
```
Comment (borisroman) [inline engine/schema/src/com/cloud/network/dao/IPAddressDaoImpl.java]: Needs to be AllocatedIpCountForDc.create()

**Comment by remibergsma** [comment]
@borisroman Please ping me if this PR is ready to be tested.

**Comment by borisroman** [comment]
@remibergsma @wilderrodrigues @karuturi Please review!

**Comment by DaanHoogland** [comment]
@borisroman there are no (unit) tests. can you add or describe or at least how to prove this fix?

**Comment by borisroman** [comment]
@DaanHoogland That was also one of my questions on Slack. There are no Unit tests regarding Dao classes in  /cloud-engine-schema/test/. So how do I test these?

I've functionally tested all by starting with a fully deployed cloud and removing all components (hosts, ips , clusters, pods, .. ) until I could confirm all exceptions where thrown in the right place.

**Comment by DaanHoogland** [comment]
@borisroman so please add a description of your manual test scheme in the description area of this PR 
( and help us think of a better dao implementation ;)

**Comment by remibergsma** [comment]
Hi @borisroman I deployed a DC and then added an extra POD. Then without creating even a cluster, tried to delete it. It failed saying:

```
The pod cannot be deleted because there are virtual machines in this pod.
```

**Comment by wilderrodrigues** [comment]
:-1:  until we get steps to test it.

Cheers,
Wilder

**Comment by remibergsma** [comment]
@borisroman Just successfully deleted a POD and a ZONE.

screen shot 2015-09-17 at 17 02 02

screen shot 2015-09-17 at 17 02 35

screen shot 2015-09-17 at 17 02 27

screen shot 2015-09-17 at 17 11 25

screen shot 2015-09-17 at 17 11 41

LGTM

**Comment by remibergsma** [comment]
For the record, I tested against fa5f388fe91bc2fd477b1f6752c2a31ac69d5051.

**Comment by remibergsma** [comment]
@wilderrodrigues The steps to test this are simply creating a POD, then deleting it. It fails in current master and succeeds in this PR. Same with a zone. If you deploy with Marvin and make sure it cannot find the hypervisor, the cleanup should succeed. That doesn't work either on current master as it leaves you with a half cleaned-up mess preventing you from deploying again.

I showed the manual tests above (screenshots). Here's the failing Marvin deploy with successful cleanup. I made sure the hypervisor was down so it couldn't succeed.

```
[root@cs1 cloudstack]# python /data/git/$HOSTNAME/cloudstack/tools/marvin/marvin/deployDataCenter.py -i /data/shared/marvin/mct-zone2-xen2.cfg

==== Log Folder Path: /tmp//MarvinLogs//DeployDataCenter__Sep_17_2015_18_20_59_Z1US54. All logs will be available here ====

==== Deploy DC Started ====
Exception Occurred :['Traceback (most recent call last):\n', '  File "/data/git/cs1/cloudstack/tools/marvin/marvin/deployDataCenter.py", line 136, in addHosts\n    ret = self.__apiClient.addHost(hostcmd)\n', '  File "/usr/lib/python2.7/site-packages/marvin/cloudstackAPI/cloudstackAPIClient.py", line 1572, in addHost\n    response = self.connection.marvinRequest(command, response_type=response, method=method)\n', '  File "/usr/lib/python2.7/site-packages/marvin/cloudstackConnection.py", line 379, in marvinRequest\n    raise e\n', 'CloudstackAPIException: Execute cmd: addhost failed, due to: errorCode: 530, errorText:Unable to add the host\n']

===deploy dc failed, so cleaning the created entries===

====DeployDC: CleanUp Started====

====Clean Up Entries=== {'PhysicalNetwork': [u'd69f38fc-6fa9-474d-a89a-b8d8a7219cf4'], 'Cluster': [u'1a51747a-fc04-4a57-9b4f-5b1b3dcaa3ba'], 'Pod': [u'73132578-a301-4e89-8397-df043568efa7'], 'order': ['Cluster', 'Pod', 'PhysicalNetwork', 'Zone'], 'Zone': [u'406ec1f9-1c90-46a1-9a3a-ee137f5baf7b']}

===Removing DataCenter Successful===
```

**Comment by karuturi** [comment]
Tested this on a xenserver 6.5 setup. delete host/storage pool/cluster/pod/network/zone are all successful. 
:+1:

**Comment by karuturi** [comment]
I am not merging this as I see a -1 from @wilderrodrigues 
@wilderrodrigues, unittests are added and steps to test are also available. Can you review again?

**Comment by wilderrodrigues** [comment]
Okay... we got 2 LGTM with proper testing/reporting. Thanks guys for taking your time to look into it.

ping @borisroman @remibergsma @karuturi @wido ... please proceed with merging.

Cheers,
Wilder

**Pull request merged**